src/out_of_order_engine.cc

```diff
@@ -0,0 +1,414 @@
+	}
+
+	// If we didn't return early so far, the instruction is eligible for eager assignment
+	assert(eagerly_assignable.has_value()); // otherwise num_incomplete_predecessors would have been == 0 above
+	node.assignment = *eagerly_assignable;
+	assignment_queue.push(node.instr);
+}
+
+void engine_impl::submit(const instruction* const instr) {
+	const auto iid = instr->get_id();
+	auto [node_it, inserted] = incomplete_instructions.emplace(iid, incomplete_instruction_state(instr));
+	assert(inserted);
+	auto& node = node_it->second;
+
+	const auto add_eligible_devices_by_memory_id = [&](const memory_id mid) {
+		/// We assume that there's either a 1:1 device <-> memory mapping for device-accessible memory, or when it's not, it's irrelevant which device we
+		/// dispatch alloc / free instructions to.
+		for(device_id did = 0; did < system.devices.size(); ++did) {
+			if(system.devices[did].native_memory == mid) { node.eligible_devices.push_back(did); }
+		}
+	};
+
+	// Perform target / queue assignment here instead of instruction graph generation time because the IDAG is too abstract of a description to know
+	// about in-order queues, and because at least for copy-instruction we want to retain dynamic queue assignment to submit to either the source or
+	// destination device and lengthen the path that can be scheduled onto a single in-order queue.
+	matchbox::match(
+	    *instr,                                //
+	    [&](const alloc_instruction& ainstr) { //
+		    node.target = target::alloc_queue;
+		    add_eligible_devices_by_memory_id(ainstr.get_allocation_id().get_memory_id());
+	    },
+	    [&](const free_instruction& finstr) { //
+		    node.target = target::alloc_queue;
+		    add_eligible_devices_by_memory_id(finstr.get_allocation_id().get_memory_id());
+	    },
+	    [&](const copy_instruction& cinstr) {
+		    const auto source_mid = cinstr.get_source_allocation().id.get_memory_id();
+		    const auto dest_mid = cinstr.get_dest_allocation().id.get_memory_id();
+
+		    // Eligible devices are tried in array-order within `assign_one`. If no other constraints exist, prefer assigning this instruction to the
+		    // dest-memory device to allow eager assignment of a subsequent kernel launch to the same in-order queue.
+		    add_eligible_devices_by_memory_id(dest_mid);
+		    add_eligible_devices_by_memory_id(source_mid);
+
+		    if(!node.eligible_devices.empty()) {
+			    node.target = target::device_queue;
+		    } else {
+			    assert(source_mid <= host_memory_id && dest_mid <= host_memory_id);
+			    node.target = target::host_queue;
+		    }
+	    },
+	    [&](const device_kernel_instruction& dkinstr) {
+		    node.target = target::device_queue;
+		    node.eligible_devices.push_back(dkinstr.get_device_id());
+	    },
+	    [&](const host_task_instruction& htinstr) { //
+		    node.target = target::host_queue;
+	    },
+	    [&](const auto& /* other */) { //
+		    node.target = target::immediate;
```

Comment:
Is it a good idea to have basically a fallthrough case here?
I'm not sure how many individual cases would have to be covered, but this way someone adding a new instruction type might overlook what happens to it in the OoO Engine.

Reply:
It's kinda ugly enumerating all these, but I agree that there should be a compile-time error when somebody forgets to update this. Fixed.